src/BaseVault.sol

```diff
@@ -0,0 +1,662 @@
+        if (asset_ == address(0)) {
+            revert ZeroAddress();
+        }
+        AssetStorage storage assetStorage = _getAssetStorage();
+        uint256 index = assetStorage.list.length;
+        if (index > 0 && assetStorage.assets[asset_].index != 0) {
+            revert DuplicateAsset(asset_);
+        }
+        assetStorage.assets[asset_] = AssetParams({active: active_, index: index, decimals: decimals_});
+        assetStorage.list.push(asset_);
+
+        emit NewAsset(asset_, decimals_, index);
+    }
+
+    /**
+     * @notice Pauses the vault.
+     */
+    function pause() external virtual onlyRole(PAUSER_ROLE) {
+        if (paused()) {
+            revert Paused();
+        }
+
+        VaultStorage storage vaultStorage = _getVaultStorage();
+        vaultStorage.paused = true;
+        emit Pause(true);
+    }
+
+    /**
+     * @notice Unpauses the vault.
+     */
+    function unpause() external virtual onlyRole(UNPAUSER_ROLE) {
+        if (!paused()) {
+            revert Unpaused();
+        }
+
+        VaultStorage storage vaultStorage = _getVaultStorage();
+        if (provider() == address(0)) {
+            revert ProviderNotSet();
+        }
+        vaultStorage.paused = false;
+        emit Pause(false);
+    }
+
+    /**
+     * @notice Processes the accounting of the vault by calculating the total base balance.
+     * @dev This function iterates through the list of assets, gets their balances and rates,
+     *      and updates the total assets denominated in the base asset.
+     */
+    function processAccounting() public virtual {
+        VaultStorage storage vaultStorage = _getVaultStorage();
+
+        uint256 totalBaseBalance = vaultStorage.countNativeAsset ? address(this).balance : 0;
+
+        AssetStorage storage assetStorage = _getAssetStorage();
+        address[] memory assetList = assetStorage.list;
+        uint256 assetListLength = assetList.length;
+        uint256 baseAssetUnit = 10 ** (assetStorage.assets[asset()].decimals);
+
+        for (uint256 i = 0; i < assetListLength; i++) {
+            uint256 balance = IERC20(assetList[i]).balanceOf(address(this));
+            if (balance == 0) continue;
+            uint256 rate = IProvider(provider()).getRate(assetList[i]);
+            totalBaseBalance += balance.mulDiv(rate, baseAssetUnit, Math.Rounding.Floor);
+        }
+
+        _getVaultStorage().totalAssets = totalBaseBalance;
+        emit ProcessAccounting(block.timestamp, totalBaseBalance);
+    }
+
+    /**
+     * @notice Processes a series of calls to target contracts.
+     * @param targets The addresses of the target contracts.
+     * @param values The values to send with the calls.
+     * @param data The calldata for the calls.
+     * @return returnData The return data from the calls.
+     */
+    function processor(address[] calldata targets, uint256[] memory values, bytes[] calldata data)
```

Comment:
seems that owners can rug the vault with this (even though `guard`). probably not ideal?